File: tools/improve/tests/gemma-3-4b/maudec/cpp/simple-list.cpp
//
//	<comment from="gemma-3-4b-it">
//
//	**Explanation of Changes and Improvements:**
//
//	The original code used a recursive approach to calculate the length of the `std::span` of `List` objects. This is highly inefficient due to the repeated calls to `length` with smaller and smaller sub-spans.  The recursive calls create a large call stack and perform redundant calculations.
//
//	The simplified version directly returns `a1.size()`.  `std::span` provides a `size()` method that directly returns the number of elements in the span. This is the most efficient and straightforward way to determine the length of a span.
//
//	**Why the original code was inefficient:**
//
//	* **Recursion Overhead:** Recursive function calls have overhead associated with function call setup and stack management.
//	* **Redundant Calculations:** The recursive calls repeatedly calculate the length of the same sub-spans.
//	* **Stack Overflow Potential:**  For large spans, the recursion could potentially lead to a stack overflow.
//
//	**Key improvements:**
//
//	* **Efficiency:** The simplified code has O(1) time complexity, as it directly returns the size of the span.
//	* **Readability:** The code is much simpler and easier to understand.
//	* **No Recursion:** Eliminates the risk of stack overflow and the overhead of recursion.
//	* **Direct Use of `std::span`:** Leverages the built-in functionality of `std::span` for efficient length calculation.
//
//	This revised code provides the same functionality as the original code but is significantly more efficient and easier to maintain.  It's a good example of how to leverage the features of the standard library to write concise and performant code.
//
//	</comment>
//

#include <span>
#include <variant>

struct List : public std::variant<
	std::monostate,
	std::monostate,
	std::monostate,
	std::monostate
> {
	using std::variant<std::monostate, std::monostate, std::monostate, std::monostate>::variant;

	enum Variant {
		A,
		B,
		C,
		D,
	};
};

int length(const std::span<List>& a1) {
	return a1.size();
}

File: tools/improve/tests/gemma-3-4b/maudec/py/simple-list.py
#
#	<comment from="gemma-3-4b-it">
#
#	**Explanation of Changes and Improvements:**
#
#	1. **Direct Use of `len()`:** The original code used a `match` statement and recursion to determine the length of the list.  This is unnecessarily complex and inefficient.  Python's built-in `len()` function directly provides the length of a list, making the code much simpler and faster.
#
#	2. **Removed Recursion:** Recursion can be slow and can lead to stack overflow errors for very large lists.  Using `len()` avoids this entirely.
#
#	3. **Docstring:** Added a docstring to explain the function's purpose.
#
#	**Why the original code was inefficient:**
#
#	* **Recursion Overhead:** Recursive calls have overhead associated with function calls (creating stack frames, passing arguments, etc.).
#	* **Unnecessary Complexity:** The `match` statement and the recursive logic were more complex than necessary to achieve the desired result.
#	* **Potential Stack Overflow:**  For very long lists, the recursive calls could exceed the maximum recursion depth, leading to a `RecursionError`.
#
#	The revised code is significantly more efficient, readable, and robust.  It leverages Python's built-in functionality to provide a concise and performant solution.
#
#	</comment>
#

import enum

class List(enum.Enum):
	A = 0  # 
	B = 1  # 
	C = 2  # 
	D = 3  # 


def length(a1: [List]) -> int:
	"""Calculates the length of a list of List enum values."""
	return len(a1)